dbms/src/Storages/MergeTree/MergeTreeReader.h

```diff
@@ -123,10 +123,15 @@ class MergeTreeReader : private boost::noncopyable
 
     void readData(
         const String & name, const IDataType & type, IColumn & column,
-        size_t from_mark, size_t max_rows_to_read,
+        size_t from_mark, bool seek_to_from_mark, size_t max_rows_to_read,
         size_t level = 0, bool read_offsets = true);
 
     void fillMissingColumnsImpl(Block & res, const Names & ordered_names, bool always_reorder);
+
+    /// return the number of rows has been read or zero if ther is no columns to read
+    size_t readRange(size_t from_mark, bool seek_to_from_mark, size_t max_rows_to_read, Block & res);
```

Comment:
As I understand, this is used to start reading range or continue reading after some part of granule was read.
And this method do it without MergeTreeRangeReader.

Probably you could invent better name or restructure code.